[Localization] Delay loading localization on startup

HassanElDesouky opened the pull request:
We were loading localization files on start-up, but now we need to stop loading them and instead load them when first diagnostic is requested, so there is no startup overhead for the compiler on projects without errors.

cc @xedin

HassanElDesouky commented on tools/libSwiftSyntaxParser/CMakeLists.txt:
```diff
@@ -14,7 +14,8 @@ if(NOT SWIFT_BUILT_STANDALONE AND NOT CMAKE_C_COMPILER_ID MATCHES Clang)
 target_link_libraries(libSwiftSyntaxParser PRIVATE
-  swiftParse)
+  swiftParse
+  swiftLocalization)
```
I was getting linker error message when building `libSwiftSyntaxParser` and `swift-dependency-tool`. I thought that was because we're not linking `swiftLocalization` here. 

After doing that change it worked. So I have two questions:
1. Is it the correct way to fix it?
2. I'm not entirely sure why it is  happening now and not before? I.e. why my change to call `setLocalization` from `DiagnosticEngine::diagnosticMessageFor` caused that linker error?

An example of the linker error:
```
[1/2] Linking CXX shared library lib/lib_InternalSwiftSyntaxParser.dylib
FAILED: lib/lib_InternalSwiftSyntaxParser.dylib
Undefined symbols for architecture x86_64:
...
...
...
"swift::diag::YAMLLocalizationProducer::YAMLLocalizationProducer(llvm::StringRef, bool)", referenced from:
      swift::DiagnosticEngine::setLocalization(llvm::StringRef, llvm::StringRef) in libswiftAST.a(DiagnosticEngine.cpp.o)
  "swift::diag::SerializedLocalizationProducer::SerializedLocalizationProducer(std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> >, bool)", referenced from:
      swift::DiagnosticEngine::setLocalization(llvm::StringRef, llvm::StringRef) in libswiftAST.a(DiagnosticEngine.cpp.o)
ld: symbol(s) not found for architecture x86_64
```

xedin replied:
It used be to called from different module before - Frontend.cpp belongs to `swiftFrontend`, I think it should stay that way.

HassanElDesouky replied:
Makes sense, thank you!

xedin commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -1129,7 +1129,14 @@ DiagnosticEngine::diagnosticStringFor(const DiagID id,
+  // there is no startup overhead for the compiler on projects without errors.
+  if (!getLocalizationLocale().empty()) {
+    if (!localization) {
+      setLocalization(getLocalizationLocale(), getLocalizationPath());
```
Here is a problem with this approach - previously if `setLocalization` where to fail it would do so only once and set a pointer to `nullptr`, but now it would have to try and fail on every diagnostic produced by the program. I'd suggest an alternative approach (in the light of linker failures) - keep `setLocalization` the way it is and call it from `Frontend` but adjust `LocalizationProducer` to avoid loading data until first request. If it fails to load YAML, let's just fallback to using diagnostic strings from that point forward.


I'd suggest we rename `setLocalization` into `initializeLocalization` which would return a `std::unique_ptr<

HassanElDesouky replied:
Do you mean we should go with the first approach I suggested which was 
- Make `YAMLLocalizationProducer::diagnostics` vector an optional.
- Remove the logic inside of the `YAMLLocalizationProducer` ’s constructor 
- In `getMessage` and `YAMLLocalizationProducer::forEachAvailable` we will check if diagnostics isn’t initialized, we will read the yaml file like we currently do in the constructor. (Move the logic of the YAMLLocalizationProducer ’s constructor to inside of this if statement)

I think since YamlProducer is only called in `setLocalization` that way we will fail once and will set `localization` pointer to nullptr. However, in other places in which we use YAMLLocalizationProducer like in tools and in unit tests... we will be trying to load the yaml file each time we try to get a message from maybe `YAMLLocalizationProducer::forEachAvailable`. So it's the same problem, as I understand... I think I will need some sort of flag so I don't try to initialize `diagnostics` again if it has failed.

I think we can overcome this with a simpler solution which is just add a boolean field to DiagnosticEngine that will check if we tried to call setLocalization or not, and in `diagnosticStringFor` we will check `if (!localization && !triedtoSetLocaliztion)`. I'm not sure however whether this approach considered "clean" or not and not sure also how I will be able to solve the linker failure without what I'm currently doing.

xedin replied:
> Do you mean we should go with the first approach I suggested which was

Note this is not only for YAML, `SerializedLocalizationProducer` shouldn't load anything either. I don't think it would be necessary to make `diagnostics` optional if there is a state recorded by `LocalizationProducer` regarding whether diagnostics have been initialized yet or not.

> I think we can overcome this with a simpler solution which is just add a boolean field to DiagnosticEngine that will check if we tried to call setLocalization or not, and in diagnosticStringFor we will check if (!localization && !triedtoSetLocaliztion). I'm not sure however whether this approach considered "clean" or not and not sure also how I will be able to solve the linker failure without what I'm currently doing.

How about we add such a flag to localization producer itself and provide a method for `DiagnosticEngine` to check whether it's valid or not?

HassanElDesouky replied:
> How about we add such a flag to localization producer itself and provide a method for `DiagnosticEngine` to check whether it's valid or not?


This is a good idea. I think we can change `LocalizationProducer` and add a `bool validStatus` field. In the constructor we will make it equal `true` with `validStatus(true)`, and add a method to use in DiagEngine.
```
  /// Check if we tried to initialize LocalizationProducer or not.
  ///
  /// \returns true if we tried to initialize LocalizationProducer before, false otherwise.
  bool isValid() const { return validStatus; }
``` 
Finally, in `diagnosticStringFor` I can check with `if (!localization && !localization.get()->isValid())`.

xedin replied:
Yeah, you can define a boolean operator on `Localization` producer so that you can simplify check down to `if (auto producer =  localization.get())`

xedin commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -162,17 +162,24 @@ class LocalizationProducer {
   LocalizationProducer(bool printDiagnosticNames = false)
       : localizationSaver(localizationAllocator),
-        printDiagnosticNames(printDiagnosticNames) {}
+        printDiagnosticNames(printDiagnosticNames), validStatus(true) {}
```
It can't be `true` until the initialization has actually attempted right?

xedin replied:
This would have to be set by localization producer implementations.

xedin reviewed (commented):
This is not exactly what we have discussed.  I still don't think there is any need to save locale/path in the `DiagnosticEngine`, I think that belongs to `LocalizationProducer` as does most of the logic contained by `DiagnosticEngine::setLocalization`.

HassanElDesouky commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -802,29 +802,11 @@ namespace swift {
-              filePath.str(), getPrintDiagnosticNames());
-        }
-      }
+      localization = diag::LocalizationProducer::producerFor(
+          locale, path, getPrintDiagnosticNames());
```
I moved the logic which was inside of `setLocalization` to the localization producer itself. I think it makes sense to add it there. WDYT?

HassanElDesouky commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -194,9 +208,10 @@ class YAMLLocalizationProducer final : public LocalizationProducer {
   /// with its unique identifier.
   void forEachAvailable(
-      llvm::function_ref<void(swift::DiagID, llvm::StringRef)> callback) const;
+      llvm::function_ref<void(swift::DiagID, llvm::StringRef)> callback);
```
I had to remove `const` because this method can initialize the  YAMLLocalizationProducer

xedin replied:
Sounds good!

HassanElDesouky commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -44,6 +44,8 @@ namespace diag {
 
 using namespace llvm::support;
 
+enum LocalizationProducerState : bool { NotInitialized, Initialized };
```
I think this is kinda more descriptive than managing a raw boolean

xedin replied:
How about we add one more state here - `FailedInitialization`?

HassanElDesouky commented on lib/Localization/LocalizationFormat.cpp:
```diff
@@ -96,6 +96,10 @@ bool SerializedLocalizationWriter::emit(llvm::StringRef filePath) {
 llvm::StringRef
 LocalizationProducer::getMessageOr(swift::DiagID id,
                                    llvm::StringRef defaultMessage) {
+  if (state == NotInitialized) {
+    initialize();
+    state = Initialized;
+  }
```
That way we will try to initialize the localization producer only once. Right?

xedin replied:
See my comment for `initialize()` declaration.

xedin replied:
Also this should check for `state == FailedInitialization` and produce default if so.

xedin commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -173,16 +175,28 @@ class LocalizationProducer {
+  /// diagnostics file available, otherwise returns a `YAMLLocalizationProducer`
+  /// if the `YAML` file is available. If both files aren't available returns a
+  /// `nullptr`.
+  static std::unique_ptr<LocalizationProducer> producerFor(llvm::StringRef locale, llvm::StringRef path, bool printDiagnosticNames);
```
Nit: needs formatting.

xedin commented on lib/Localization/LocalizationFormat.cpp:
```diff
@@ -141,14 +153,44 @@ llvm::StringRef YAMLLocalizationProducer::getMessage(swift::DiagID id) const {
   }
 }
 
+std::unique_ptr<LocalizationProducer> LocalizationProducer::producerFor(llvm::StringRef locale, llvm::StringRef path, bool printDiagnosticNames) {
```
Nit: needs formatting.

xedin commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -173,16 +175,28 @@ class LocalizationProducer {
   virtual ~LocalizationProducer() {}
 
 protected:
+  LocalizationProducerState state;
```
This has to be initialized to `false`, either directly here or in a constructor of `LocalizationProducer` and it should not be visible to subtypes (made `private`) because that would make it possible to mess with that state.

xedin commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -173,16 +175,28 @@ class LocalizationProducer {
+  LocalizationProducerState state;
+
+  /// Used to lazily initialize `LocalizationProducer`
+  virtual void initialize() = 0;
```
I suggest you add a `virtual bool initializeImpl() = 0;` (return indicates whether initialization succeeded or failed) and change this one to:

```
virtual void initialize() final {
   if (initializeImpl())
    state = Initialized;
   else
    state = FailedInitialization;
}
```

This way `LocalizationProducer` controls initialization done by sub-classes and `state` flag is protected.

HassanElDesouky replied:
I think a better way to do that would be to add a check in `initialize` to check whether the localization producer has been initialized. I.e. 
```
virtual void initialize() final {
  if (state == NotInitialized && initializeImpl())
    state = Initialized;
  else
    state = FailedInitialization;
}
```
That way we won't proceed to `initializeImpl()` unless we are sure that the localization producer hasn't been initialized before. 

By making that change, I think we can simplify the checking statement in `LocalizationProducer::getMessageOr` to:
```
LocalizationProducer::getMessageOr(swift::DiagID id,
                                   llvm::StringRef defaultMessage) {
  initialize();
  if (state == FailedInitialization) {
    return defaultMessage;
  }
...
...
```
and in `YAMLLocalizationProducer::forEachAvailable` it will also be:
```
void YAMLLocalizationProducer::forEachAvailable(
    llvm::function_ref<void(swift::DiagID, llvm::StringRef)> callback) {
  initialize();
...
...
...
```

xedin replied:
Sure, rename initialize to initializeIfNeeded then.

xedin replied:
Actually renaming that allows us to switch back to `initialize()` for implementation specific one, also:

```
if (state != NotInitialized)
  return;

if (initializeImpl())
  ...
else
  ...
```

Is better because otherwise `state` would be constantly reset for no reason.

xedin commented on lib/Localization/LocalizationFormat.cpp:
```diff
@@ -141,14 +153,44 @@ llvm::StringRef YAMLLocalizationProducer::getMessage(swift::DiagID id) const {
+    llvm::function_ref<void(swift::DiagID, llvm::StringRef)> callback) {
+  if (state == NotInitialized) {
+    initialize();
+    state = Initialized;
```
Should check for `state == FailedInitialization` as well.

xedin reviewed (changes requested):
Looks much better now, I have left some more comments inline.

HassanElDesouky commented on include/swift/Localization/LocalizationFormat.h:
```diff
@@ -173,16 +180,33 @@ class LocalizationProducer {
+  /// Used to lazily initialize `LocalizationProducer`s.
+  /// \returns true if the producer is successfully initialized, false
+  /// otherwise.
+  virtual bool initializeImpl() = 0;
```
@xedin I'm just not sure when `YAMLLocalizationProducer::initializeImpl` or `SerializedLocalizationProducer::initializeImpl` will ever fail and return false. I'm returning true in both implementations for now.

HassanElDesouky replied:
I mean, the current state (before this PR) is that in both Serialized & YAML LocalizationProducers' constructors doesn't fail. Or at least we don't handle failures in the constructors.

xedin replied:
It might not matter now but APIs change so it's prudent to account for possibility of a failure.

HassanElDesouky replied:
Makes sense, thank you!

HassanElDesouky committed:
[Localization] Delay loading localization

- Add `LocalizationProducerState` to manage the states of LocalizationProducers.
- Add `initializeImpl` and `initializeIfNeeded` to manage lazily initialization of `LocalizationProducer`s.
- Move constructing a localization producer from DiagEngine to `LocalizationProrducer` itself.

xedin reviewed (approved):
Great, thank you!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please clean test Linux platform

xedin commented:
@swift-ci please test Linux platform
